src/hooks/useDataErrorHandler.js

```diff
@@ -0,0 +1,26 @@
+import { useEffect, useState, useRef } from 'react';
+
+export const useDataErrorHandler = (callback, { invalidSegments }) => {
+  const [dataError, setDataError] = useState({});
+  const callbackRef = useRef(callback);
+
+  useEffect(() => {
+    callbackRef.current = callback;
+  }, [callback]);
+
+  useEffect(() => {
+    setDataError(prevErrors => {
+      const { ...errorsWithoutSegments } = prevErrors;
+      delete errorsWithoutSegments.segments;
+      return invalidSegments.length ? { ...errorsWithoutSegments, segments: invalidSegments } : errorsWithoutSegments;
+    });
+  }, [invalidSegments]);
+
+  useEffect(() => {
+    if (Object.keys(dataError).length > 0 && callbackRef.current) {
+      callbackRef.current(dataError);
+    }
+  }, [dataError]);
```

Comment:
I was wondering something...  Any reasons why this simplified implementation would not be sufficient? 

```suggestion
  useEffect(() => {
    if (Object.keys(invalidSegments).length > 0 && callback) {
      callback(invalidSegments);
    }
  }, [invalidSegments]);

```

Reply:
1. The `callback` instance being used is the first one, not the latest, because it's not included in the useEffect dependency array. Including it in the dependencies would cause multiple warnings, especially if the callback is defined as anonymous, such as `onDataError={() => {...}}`.
2. While I could create a separate hook to calculate the errors, the current solution returns a dataError, which I then use to conditionally display the error component below the arc.
3. It can have more than one error in this lib, not only segments, so it would need to be at least pass `callback({segments: invalidSegments});`

Reply:
Do you mean callback function can be dynamic that's why we want to watch it? I can't think of instances of it being changed after the component changes. We don't need to watch it in the array either.

Reply:
Considering the `callback` is used only when `invalidSegments` props are updated, I believe there'll be no bug removing the useRef besides breaking the react rule. So, I'll update this, as I said, I need the `dataError` object to defined If I can show/hide the `dataErrorComponent`

//useDataErrorHandler.js
```js
import { useEffect } from 'react';

const useDataErrorHandler = (callback, dataError) => {
  useEffect(() => {
    if (callback && Object.keys(dataError).length > 0) {
      callback(dataError);
    }
  }, [dataError]);
};

export default useDataErrorHandler;


```

// useDataError.js
```js
import { useEffect, useState } from 'react';

const useDataError = ({ invalidSegments }) => {
  const [dataError, setDataError] = useState({});

  useEffect(() => {
    setDataError(prevErrors => {
      const { ...errorsWithoutSegments } = prevErrors;
      delete errorsWithoutSegments.segments;
      return invalidSegments.length ? { ...errorsWithoutSegments, segments: invalidSegments } : errorsWithoutSegments;
    });
  }, [invalidSegments]);

  return dataError;
};

export default useDataError;

```

Reply:
Why do you need useDataError if you can directly use `invalidSegments`?

Reply:
If I don't have a centralized place to detect data errors, I need to have the same business logic `invalidSegments.length > 0` in different places.

```js
import { useEffect } from 'react';

const useDataErrorCallback = (callback, {invalidSegments}) => {
  useEffect(() => {
    if (callback && invalidSegments.length > 0) {
      callback({segments: invalidSegments});
    }
  }, [dataError]);};

export default useDataErrorCallback;
```

and


 ```js
const hasInvalidProps =invalidSegments.length > 0
```

Now, let’s say I need to handle invalid props for `fillValue` and `arcDegree` in the next PR, as they are also causing crashes. The issue is that I would need to duplicate the error-handling logic in two separate places. If someone adds new error handling to the `useDataErrorCallback` but forgets to include it in the `const hasInvalidProps =invalidSegments.length > 0`, it could introduce a new bug.

Reply:
That's another alternative,  create a function like this to avoid needing another hook. 

Some utils
```js
export const getErrors = ({ invalidSegments }) => {
  let dataError = {};
  if (invalidSegments.length > 0) {
    dataError.segments = invalidSegments;
  }

  return dataError;
};
```

SegmentArc
```js
  const dataError = useMemo(() => getErrors({ invalidSegments }), [invalidSegments]);
  useDataErrorCallback(onDataError, dataError);
```
The difference is that it would need a useMemo, to not activate the `useDataErrorCallback` useEffect multiple times because of internal lib state change:
useDataErrorCallback
```js
import { useEffect } from 'react';

const useDataErrorCallback = (callback, dataError) => {
  useEffect(() => {
    if (callback && Object.keys(dataError).length > 0) {
      callback(dataError);
    }
  }, [dataError]);
};

export default useDataErrorCallback;
```

Reply:
I think we can keep the original solution. It is best to follow React's practices to avoid additional questions or proposals to change it.